### DISCUSSION TITLE
Config Injection

### DISCUSSION THREAD

**Discussion by Emily-Jiang** [eclipse-ee4j/config#126] [Ideas]
For config injection, the dependency will be on JSR 330. 

Propose :
- Support a singlar property injection for single property lookup e.g. `@Inject @ConfigProperty("name") String/other type` 
- Support a type injection e.g. `@Inject @ConfigProperty("prefix") MyConfigMapping mapping;`

The property lookup will be identical for single property or ConfigMapping. For ConfigMapping, the value will be treated as prefix while for non configmapping, the name will be the exact property name.

Any thoughts on this?

**Comment by dmlloyd** [comment]
I would vote "no" on single property injection.

For type injection of configuration mapped objects, I would suggest only supporting the simple form for MVP: `@Inject MyConfigMapping mapping;`.

**Reply by Emily-Jiang** [reply]
What about supplying prefix to injection?
```
client.name=Bob
client.age=23
customer.name=Alice
customer.age=33
```
We have MyConfigMapping to contain name and age.
When we do the injection, if they want to get `client` info so that they can specify the injection with the prefix with `client`.

**Reply by m0mus** [reply]
I agree with David. Prefix can be specified as an optional parameter of the annotation placed on the interface.

**Reply by dmlloyd** [reply]
Here is an example:

```java
public interface MyConfigMapping {
    String name();
    int age();
}

@ConfigMapping // ← this is only (?) strictly necessary when injecting so maybe it should be part of that instead; maybe call it `@InjectableConfig` or something...
public interface OuterMapping {
    MyConfigMapping customer();
    MyConfigMapping client();
}

public class MyBean {
    @Inject OuterMapping theMapping;

    // …
}
```

**Reply by Emily-Jiang** [reply]
David and I discussed this during today's call and agreed. In the MVP, the prefix will not be configured. In this case, the annotation `ConfigMapping` is not required as only one bean is available.

**Reply by Emily-Jiang** [reply]
I will create a PR with a few sentences for further discussion.

**Reply by radcortez** [reply]
The injection point should not require (or override the prefix). I believe that was a mistake we made in MP Config. It is a pain to maintain and implement, and it hardly has any usage.

But we still need a way to set up the initial path segment to start the mapping process.
